### PR TITLE
Refactored sym_expr to fix documentation problems with SymPyExpression

### LINKED ISSUE #18219: SymPyExpression documentation not rendering
The documentation for SymPy Expressions is not rendering in https://docs.sympy.org/latest/modules/parsing.html as it uses 
```python
class SymPyExpression(object):
        def __init__(self, *args, **kwargs):
            raise ImportError('Module not available.')
```
as the source and not 
https://github.com/sympy/sympy/blob/7f92c96417d066700d096105282d6fe47ea085ef/sympy/parsing/sym_expr.py#L20 

Also, I think I should remove the Fortran Parsing Reference from the docs as it is a backend implementation and the users will not interact with the Module.

### PR DESCRIPTION
**Pull request by Sc0rpi0n101**
#### References to other Issues or PRs

Fixes #18219 

#### Brief description of what is fixed or changed
Modified sym_expr to remove the dummy SymPyExpression class and moved the actual SymPyExpression class to the top level.

Removed the Fortran Parsing Reference from the documentation as it is the backend implementation and the end users would not interact with it. The developers can still check out the required documentation in the docstrings.

 Also updated the installation instructions to include clang and LFortran installation instructions.

#### Other comments

Should I add `SymPyExpression` to the `__init__.py` file for the parsing module?

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by Sc0rpi0n101** [comment]
Here are some images of updated documentation for reference:

image

image

File: sympy/parsing/sym_expr.py [hunk 1]
```diff
@@ -5,281 +5,275 @@
...
-    class SymPyExpression(object):
-        def __init__(self, *args, **kwargs):
-            raise ImportError('Module not available.')
+if lfortran:
+    from sympy.parsing.fortran.fortran_parser import src_to_sympy
+if cin:
+    from sympy.parsing.c.c_parser import parse_c
+
+@doctest_depends_on(modules=['lfortran', 'clang.cindex'])
+class SymPyExpression(object):  # type: ignore
```
Comment (oscarbenjamin) [inline sympy/parsing/sym_expr.py]: Is the `# type: ignore` still needed? I think it was just there because of the duplicate class

**Comment by oscarbenjamin** [comment]
> Should I add `SymPyExpression` to the `__init__.py` file for the parsing module?

The documentation says it is experimental. Is that still the case? I find the name `SymPyExpression` confusing since a normal SymPy expression is an Expr and this is something else.

**Comment by smitgajjar** [comment]
Suggestion: rename the class to `Parse` to show that we will parse given source code to sympy expression!

**Comment by Sc0rpi0n101** [comment]
> The documentation says it is experimental. Is that still the case? I find the name `SymPyExpression` confusing since a normal SymPy expression is an Expr and this is something else.

Currently, SymPyExpression is able to represent and manage the parsers effectively. But, if someone finds a better alternative during this GSoC, it can be changed. So, I'd still consider it experimental for now. 

And it is not `Expr`. It's basically a list of AST Nodes. It named it like this as it represents SymPy's Expressions. 

> Suggestion: rename the class to `Parse` to show that we will parse given source code to sympy expression!

Good suggestion. Are you sure `Parse` is not used anywhere else in the parsing module?

**Comment by smitgajjar** [comment]
Yes, it is not used elsewhere(used `grep -r "Parse("`) (sorry for the late response, I was unaware of this thread, maybe I unsubscribed accidentally!)

**Comment by Sc0rpi0n101** [comment]
> Yes, it is not used elsewhere(used `grep -r "Parse("`) (sorry for the late response, I was unaware of this thread, maybe I unsubscribed accidentally!)

Great, We can discuss that during GSoC if we can improve the API or find a better alternative.

**Comment by oscarbenjamin** [comment]
Looks good

**Pull request merged**